rfcs/0132-meson-builds-nix.md

```diff
@@ -0,0 +1,198 @@
+
+A new build system, and any build system for that matter, indirectly brings its own dependencies to the Nix project.
+
+Ideally, such dependencies should be minimal, both in number and complexity, with extra good points for dependencies already present (e.g the ubiquitous C compiler).
+
+About this specific point, a non-negligible drawback is: the reference implementation of Meson is written in Python. At least theoretically, it brings the necessity of including a Python 3 interpreter on the _bootstrap route_ of Nix.
+
+However, some points can be laid out on the opposite side:
+
+1. By design, the meson reference evaluator depends only on Python, avoiding the use of extra libraries.
+2. Also by design, the implementation language is never exposed to the meson DSL. Because of it, the reimplementation of Meson in other programming languages becomes way more facilitated.
+   1. Indeed, Muon is an alternative implementation of Meson written in C.
+3. As part of this evaluation of this bootstrap route, we should also evaluate the current bootstrap route, in order to have a fair comparison.
+
+Further, in principle the same criticisms and answers can be laid out for Ninja; however, Ninja is written in C++, a language already used to implement Nix. Therefore the bootstrap route suffers little to no alteration here.
+
+## Learning the new system
+
+A somewhat subjective but non-negligible issue is the entry barrier of this new build system. Switching from a known build system to one unknown is not without its problems.
+
+However, the Meson development team strives to keep the DSL easy to learn and pleasurable to use. It should not be hard to become familiar with the Python-esque syntax of meson, and its functional, stateless approach is certainly a feature highly appreciated by the Nix community as a whole.
+
+The huge advantages of implementing Meson surpass the small drawbacks of learning it.
+
+## Source code changes
+
+Further to the obvious inclusion of meson files (and the removal of the old quasi-autotools ones), there is a reasonable expectation of code refactoring.
+
+However, such refactorings are completely validated on the long term goals of Nix, in particular the improvements on portability.
+
+## End users
+
+The most known end user of Nix is certainly Nixpkgs. However, there are many other Linux distributions that already keep Nix on their repositories (15 families, according to Repology). There is also a reasonable expectation of affecting those package managers' devteams.
+
+However, most (if not all) of those distributions already have Meson and its companion tool Ninja in their respective package databases (53 families, according to Repology), given that many open source projects use them as build system.
+
+## Transition between old and new build infrastructure
+
+The transition between between the old and new build systems should be as smooth and controlled as possible.
+
+# Alternatives
+[alternatives]: #alternatives
```

Comment:
I believe it is well summarized on the examples of waf and premake.

Reply:
Not exactly. The argument against them is that they are a language-specific DSL, but that's not the case of Bazel or Buck.

They do have a bootstrap problem ofc, but it's a different one (single implem, lots of deps)

Reply:
A different one, and considerably worse than Meson. Cmake would be way better than them.

I do not believe they deserve more than a mention. If anyone is interested on picking Buck or Bazel or Imake, kudos to them!

Reply:
Well now, let's be fair, there's definitely more than one implementation of Bazel.

The original design of it is Google's internal closed-source thing designed specifically to meet Google's distributed cloud computing needs, after all. ;) Bazel itself is just an open-source attempt to make `BUILD` / `*.bzl` files look like an acceptable thing for other projects to include to make Google's life easier. ;)

Reply:
i think it's fair to say that anyone who has used scons/bazel/cmake v1.x would rather implode than return to the dark ages

Reply:
@eli-schwartz you are correct; but the minor improvements to bazel doesn't address its fundamental issues. plus, with googles track record of just randomly killing technology without warning, i dont think it would be wise to tie yourself to a product where they are one of the primary supporters/influencers

Reply:
I was being sarcastic. :)

Reply:
I never said that that they were a good choice (I happen to somewhat like Bazel on a lot of aspects, but I totally agree that it would be a terrible choice for Nix :) ).

But I've seen more people advocating for building Nix with Bazel (less buck because it's more niche) than all the alternatives combined (meson excepted). So just waving it out of the discussion without explaining why is probably not a great idea

Reply:
fair point; @AndersonTorres maybe the RFC should include a section "why not bazel" like it does for cmake?

Reply:
Yeah I think that is a good idea. Meson's subproject support (something CMake doesn't have, incidentally) can be contrasted with the "monorepo vibes" of Bazel/Buck as a sort of best of both worlds.

Reply:
My list of complaints:

1. I know nothing about them
2. They are built in Java
3. There is no alternative implementation in another language like C or C++ or even Python
4. I do not like Java
5. Java support on Nixpkgs is not the best in this planet
6. [Bootstrap Java is a hell](https://bootstrappable.org/projects/java.html):

   > Unfortunately, most of the software needed for the bootstrap has been abandoned. To ensure that the JDK can be built from sources without the need for an existing installation of the OpenJDK we propose to continue maintaining the links of this bootstrap chain.

7. Java is heavier than Python
8. Have I said I do not like Java?

Reply:
yep sounds about right 🤣

Reply:
Maybe too late. Has Zig been considered? It [claims being an alternative](https://ziglang.org/learn/why_zig_rust_d_cpp/#:~:text=within%20the%20implementations.-,A%20Package%20Manager%20and%20Build%20System%20for%20Existing%20Projects,-Zig%20is%20a):

> but you can use Zig as a replacement for autotools, cmake, make, scons, ninja, etc.

> Zig is offering to replace a project’s build system with a reasonable language using a declarative API for building projects, that also provides package management, and thus the ability to actually depend on other C libraries. The ability to have dependencies enables higher level abstractions, and thus the proliferation of reusable high-level code.

Reply:
the entire point of meson is it offers the ability to compile *all* projects regardless of source language, build environment, linker, or compiler configurations. it offers flexibility and simplicity which allows easy maintainability into the future. the zig buildsystem is designed for c/c++ and is not nearly as well documented or flexible.

Reply:
Also Zig is oriented to Zig itself, while Meson is general-purpose.
Using Zig as Autotools replacement would be as worse as - or even worser than - using Bazel or Waf.